### PR TITLE
Add TLS end to end unit test

### PR DESCRIPTION
**Pull request by Arkatufus**

### CODE REVIEW

File: src/TurboMqtt/Client/ClientTlsOptions.cs [hunk 1]
```diff
@@ -18,4 +18,8 @@ public sealed record ClientTlsOptions
     public bool UseTls { get; init; } = false;
     
     public SslClientAuthenticationOptions? SslOptions { get; init; }
+    
+    public RemoteCertificateValidationCallback? ServerCertificateValidator { get; set; }
```
Comment (Aaronontheweb) [inline src/TurboMqtt/Client/ClientTlsOptions.cs]: These are both built into `SslClientAuthenticationOptions` https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslclientauthenticationoptions?view=net-8.0#properties - let's use those properties instead

File: src/TurboMqtt/Client/IMqttClientFactory.cs [hunk 2]
```diff
@@ -53,6 +53,9 @@ public MqttClientFactory(ActorSystem system)
     public async Task<IMqttClient> CreateTcpClient(MqttClientConnectOptions options, MqttClientTcpOptions tcpOptions)
     {
         AssertMqtt311(options);
+        if (tcpOptions.TlsOptions is { UseTls: true, SslOptions: null })
+            throw new NullReferenceException("TlsOptions.SslOptions can not be null if TlsOptions.UseTls is true");
```
Comment (Aaronontheweb) [inline src/TurboMqtt/Client/IMqttClientFactory.cs:57]: LGTM

File: src/TurboMqtt/IO/Tcp/FakeMqttTcpServer.cs [hunk 3]
```diff
@@ -26,18 +26,22 @@ public MqttTcpServerOptions(string host, int port)
...
-    public string Host { get; }
+    public string Host { get; init; }
 
-    public int Port { get; }
+    public int Port { get; init; }
 
-    public SslServerAuthenticationOptions? SslOptions { get; set; }
+    public SslServerAuthenticationOptions? SslOptions { get; init; }
+    
+    public RemoteCertificateValidationCallback? ClientCertificateValidator { get; init; }
```
Comment (Aaronontheweb) [inline src/TurboMqtt/IO/Tcp/FakeMqttTcpServer.cs]: Same feedback as earlier - `SslServerAuthenticationOptions` also contains these properties: https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslserverauthenticationoptions?view=net-8.0 - let's use those

File: tests/TurboMqtt.Tests/TurboMqtt.Tests.csproj
Comment (Aaronontheweb) [inline tests/TurboMqtt.Tests/TurboMqtt.Tests.csproj:1]: LGTM

File: tests/certs/server.csr [hunk 4]
```diff
@@ -0,0 +1,17 @@
+-----BEGIN CERTIFICATE REQUEST-----
```
Comment (Aaronontheweb) [inline tests/certs/server.csr:1]: Did my old certs not work correctly?

File: tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+
+    protected override void AfterAll()
+    {
+        // shut down our local TCP server
+        _server.Shutdown();
+        base.AfterAll();
+    }
+
+    [Fact]
+    public async Task ShouldAutomaticallyReconnectAndSubscribeAfterServerDisconnect()
```
Comment (Aaronontheweb) [inline tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs]: Would it be better if we just made the TCP end 2 end specs sub-classable and reuse those @Arkatufus ?

**Review by Aaronontheweb — CHANGES_REQUESTED** [review]
Some very minor feedback but otherwise LGTM

File: tests/certs/server.csr (hunk 4 above)
Comment (Arkatufus) [reply tests/certs/server.csr:1]: I also needed the self-signed root CA certificate that signs the server certificate, that's why I needed to rebuild all of the certs

File: tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs (hunk 5 above)
Comment (Arkatufus) [reply tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs]: I can give that a try

File: src/TurboMqtt/Client/ClientTlsOptions.cs (hunk 1 above)
Comment (Arkatufus) [reply src/TurboMqtt/Client/ClientTlsOptions.cs]: Done

File: src/TurboMqtt/IO/Tcp/FakeMqttTcpServer.cs (hunk 3 above)
Comment (Arkatufus) [reply src/TurboMqtt/IO/Tcp/FakeMqttTcpServer.cs]: Done

File: tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs (hunk 5 above)
Comment (Arkatufus) [reply tests/TurboMqtt.Tests/End2End/MQTT_311/TlsMqtt311End2EndSpecs.cs]: Done

**Review by Aaronontheweb — APPROVED** [review]

**Pull request merged**